### PR TITLE
Enable Normal Mod VDS Conntrack

### PR DESCRIPTION
**Pull request by cl1111**

### CODE REVIEW

File: pkg/agent/datapath/localBridge.go [hunk 1]
```diff
@@ -52,6 +52,8 @@ type LocalBridge struct {
 	fromLocalRedirectTable         *ofctrl.Table // Table 15
 	fromLocalArpPassTable          *ofctrl.Table // Table 20
 	fromLocalArpSendToCtrlTable    *ofctrl.Table // Table 25
+	ctStateTable                   *ofctrl.Table // Table 1
+	ctCommitTable                  *ofctrl.Table // Table 70
```
Comment (KeyongSun) [inline pkg/agent/datapath/localBridge.go:56]: table 的名称可以标记一下是 normal monitor CT 的 table

File: pkg/exporter/localBridge.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
+package exporter
```
Comment (KeyongSun) [inline pkg/exporter/localBridge.go:1]: Copyright
这个文件名称建议不用 localBridge，singleBridge

File: pkg/exporter/config.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
+package exporter
```
Comment (KeyongSun) [inline pkg/exporter/config.go:1]: Copyright

File: pkg/exporter/localBridge.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+// Controller received a packet from the switch
+func (n *NormalBridge) PacketRcvd(sw *ofctrl.OFSwitch, pkt *ofctrl.PacketIn) {
+
+}
+
+// Controller received a multi-part reply from the switch
+func (n *NormalBridge) MultipartReply(sw *ofctrl.OFSwitch, rep *openflow13.MultipartReply) {
+
+}
```
Comment (KeyongSun) [inline pkg/exporter/localBridge.go:99]: 这些没用的代码可以删掉

File: pkg/exporter/localBridge.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+	log "github.com/Sirupsen/logrus"
+	"github.com/contiv/libOpenflow/openflow13"
+	"github.com/contiv/ofnet/ofctrl"
+	"github.com/contiv/ofnet/ovsdbDriver"
+	"sync"
+	"time"
+)
+
+const (
+	INPUT_TABLE     = 0
```
Comment (KeyongSun) [inline pkg/exporter/localBridge.go:14]: 这个在哪里用了吗？

File: pkg/exporter/ovs.go [hunk 6]
```diff
@@ -152,9 +180,14 @@ func (m *OvsMonitor) handleOvsUpdates(updates ovsdb.TableUpdates) {
 					m.collectorCache.AddIface(generateInterface(row.New))
 				}
 				if table == "Bridge" {
-					if FilterBridge(row.New.Fields["name"]) {
+					if FilterBridge(row.New.Fields["name"], datapath.POLICY_BRIDGE_KEYWORD) ||
+						FilterBridge(row.New.Fields["name"], datapath.CLS_BRIDGE_KEYWORD) {
```
Comment (echkenluo) [inline pkg/exporter/ovs.go:184]: Does the uplinkBridge need to be filtered ?

File: go.mod [hunk 7]
```diff
@@ -58,7 +58,8 @@ replace (
 	github.com/contiv/libOpenflow => github.com/everoute/libOpenflow v0.0.0-20210716071814-4ef09249fae5
 	github.com/contiv/libovsdb => github.com/everoute/libovsdb v0.0.0-20210326110222-6c508538aa65
 	//github.com/contiv/ofnet => github.com/everoute/ofnet v0.0.0-20211229083707-5cb1e875a38b
-	github.com/contiv/ofnet => github.com/echkenluo/ofnet v0.0.0-20220316073105-f24ad3ca4c99
+	//github.com/contiv/ofnet => github.com/echkenluo/ofnet v0.0.0-20220316073105-f24ad3ca4c99
+	github.com/contiv/ofnet => github.com/echkenluo/ofnet v0.0.0-20220329095940-b05eed62ea96
```
Comment (echkenluo) [inline go.mod:62]: This is my personal repo, we need to pick this commit to public repo in everoute dir.

File: pkg/exporter/ovs.go [hunk 8]
```diff
@@ -152,9 +180,14 @@ func (m *OvsMonitor) handleOvsUpdates(updates ovsdb.TableUpdates) {
...
 						continue
 					}
+					if !FilterBridge(row.New.Fields["name"], INTERNAL_BRIDGE_KEYWORD) &&
+						FilterNormalBridge(row.New.Fields["name"]) && reflect.DeepEqual(row.Old, empty) {
```
Comment (wu-cl) [inline pkg/exporter/ovs.go:188]: If we transfer bridge mode from chain to single, row.Old will not be empty.

File: pkg/agent/datapath/localBridge.go [hunk 9]
```diff
@@ -239,6 +241,20 @@ func (l *LocalBridge) BridgeInit() {
...
+	if err := sw.DeleteTable(CT_STATE_TABLE); err != nil {
+		log.Fatalf("Failed to delete single local bridge ctState table, error: %v", err)
+	}
+
+	if err := sw.DeleteTable(CT_COMMIT_TABLE); err != nil {
+		log.Fatalf("Failed to delete single local bridge ctCommit table, error: %v", err)
+	}
 
 	if err := l.initVlanInputTable(sw); err != nil {
 		log.Fatalf("Failed to init local bridge vlanInput table, error: %v", err)
```
Comment (wu-cl) [inline pkg/agent/datapath/localBridge.go:260]: We will lose connectivity if drop the whole default table until L259 ( initialize new vlanInputTable ). As @echkenluo  says, if something fatal, default table 0 may be empty. We would better delete ct flows accurately.

**Comment by wu-cl** [comment]
Do not commit with same names, or squash them before merge.

File: pkg/agent/datapath/config.go [hunk 10]
```diff
@@ -81,15 +79,19 @@ func getDatapathConfig() (*datapath.Config, error) {
...
+
+	nodeName, _ := os.Hostname()
+	nodeName = strings.ToLower(nodeName)
+	agentInfo.NodeName = nodeName
 
 	node := corev1.Node{}
+
 	if err := k8sClient.Get(context.Background(), client.ObjectKey{
-		Name: agentInfo.NodeName,
+		Name: nodeName,
```
Comment (wu-cl) [inline pkg/agent/datapath/config.go:94]: Why change this part?

File: pkg/agent/datapath/multiBridgeDatapath.go [hunk 11]
```diff
@@ -58,6 +58,11 @@ const (
 	FLOW_MATCH_OFFSET                   = 3
 )
 
+const (
+	SingleVdsInputHighPriority   = 100
+	SingleVdsInputNormalPriority = 5
```
Comment (wu-cl) [inline pkg/agent/datapath/multiBridgeDatapath.go:63]: single bridge

File: pkg/agent/datapath/localBridge.go (hunk 1 above)
Comment (cl1111) [reply pkg/agent/datapath/localBridge.go]: datapath/localBridge中不需要定义这两个table字段，已删除；exporter/singleBridge中定义的table加上了singleBridge前缀

File: pkg/exporter/ovs.go [hunk 12]
```diff
@@ -104,6 +104,23 @@ func FilterBridge(brName interface{}) bool {
...
+func FilterNormalBridge(brName interface{}) bool {
+	datapathConfig, err := datapath.GetDatapathConfig()
+	if err != nil {
+		klog.Fatalf("Failed to get datapath config. error: %v. ", err)
+	}
+	name := brName.(string)
+	for _, ovsbrname := range datapathConfig.ManagedVDSMap {
+		policyBrName := ovsbrname + "policy"
+		clsBrName := ovsbrname + "cls"
+		uplinkBrName := ovsbrname + "uplink"
```
Comment (wu-cl) [inline pkg/exporter/ovs.go:116]: use constant keyword

File: pkg/exporter/config.go (hunk 3 above)
Comment (cl1111) [reply pkg/exporter/config.go]: 已添加

File: pkg/exporter/ovs.go [hunk 13]
```diff
@@ -41,8 +41,9 @@ const (
 
 	SFlowPort = 6666
 
-	emptyUUID    = "00000000-0000-0000-0000-000000000000"
-	vSwitchdPath = "/var/run/openvswitch/ovs-vswitchd"
+	emptyUUID             = "00000000-0000-0000-0000-000000000000"
+	vSwitchdPath          = "/var/run/openvswitch/ovs-vswitchd"
+	InternalBridgeKeyword = "internal"
```
Comment (wu-cl) [inline pkg/exporter/ovs.go:46]: Internal bridge has fixed bridge name, not only suffix.

File: pkg/exporter/localBridge.go (hunk 4 above)
Comment (cl1111) [reply pkg/exporter/localBridge.go]: 增加这两个是为了将bridge构造为ofctl的AppInterface接口类型

File: pkg/exporter/localBridge.go (hunk 5 above)
Comment (cl1111) [reply pkg/exporter/localBridge.go]: 构造InputTable时，使用了ofswitch的DefaultTable()，不需要指定INPUT_TABLE ID，未使用INPUT_TABLE，已删除

File: pkg/exporter/ovs.go (hunk 6 above)
Comment (cl1111) [reply pkg/exporter/ovs.go:173]: uplinkBridge is filtered in following function FilterSingleBridge()

File: pkg/exporter/config.go [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+import (
+	"fmt"
```
[the remark this answers is not in the document]
Comment (cl1111) [reply pkg/exporter/config.go]: merge operation has been performed

File: pkg/agent/datapath/localBridge.go [hunk 15]
```diff
@@ -239,6 +241,20 @@ func (l *LocalBridge) BridgeInit() {
 	l.fromLocalRedirectTable, _ = sw.NewTable(FROM_LOCAL_REDIRECT_TABLE)
 	l.fromLocalArpPassTable, _ = sw.NewTable(FROM_LOCAL_ARP_PASS_TABLE)
 	l.fromLocalArpSendToCtrlTable, _ = sw.NewTable(FROM_LOCAL_ARP_TO_CONTROLLER_TABLE)
+	l.ctStateTable, _ = sw.NewTable(CT_STATE_TABLE)
+	l.ctCommitTable, _ = sw.NewTable(CT_COMMIT_TABLE)
+
+	if err := sw.DeleteTable(INPUT_TABLE); err != nil {
```
[the remark this answers is not in the document]
Comment (cl1111) [reply pkg/agent/datapath/localBridge.go]: when the localBridge does not have a normal CT table, the deletion operation will not lead to crash

File: go.mod (hunk 7 above)
Comment (cl1111) [reply go.mod]: has been modified

File: pkg/exporter/ovs.go (hunk 8 above)
Comment (cl1111) [reply pkg/exporter/ovs.go]: this filter condition has been removed

File: pkg/agent/datapath/localBridge.go (hunk 9 above)
Comment (cl1111) [reply pkg/agent/datapath/localBridge.go]: has been modified to delete the accurate flows with the specified priority

File: pkg/agent/datapath/config.go (hunk 10 above)
Comment (cl1111) [reply pkg/agent/datapath/config.go:94]: this is a misoperation and has been modified back

File: pkg/agent/datapath/multiBridgeDatapath.go (hunk 11 above)
Comment (cl1111) [reply pkg/agent/datapath/multiBridgeDatapath.go:63]: has been modified

File: pkg/exporter/ovs.go (hunk 12 above)
Comment (cl1111) [reply pkg/exporter/ovs.go:116]: has been modified

File: pkg/exporter/ovs.go (hunk 13 above)
Comment (cl1111) [reply pkg/exporter/ovs.go]: has been modified

File: pkg/exporter/ovs.go [hunk 16]
```diff
@@ -104,6 +107,26 @@ func FilterBridge(brName interface{}) bool {
...
+		klog.Fatalf("Failed to get datapath config. error: %v. ", err)
+	}
+	name := brName.(string)
+	if name == internalBridgeKeyword {
+		return false
+	}
+	for _, ovsbrname := range datapathConfig.ManagedVDSMap {
+		policyBrName := ovsbrname + policyKeyWord
+		clsBrName := ovsbrname + clsKeyWord
+		uplinkBrName := ovsbrname + uplinkKeyWord
```
Comment (KeyongSun) [inline pkg/exporter/ovs.go:122]: "ovsbr-mgt-policy" 这个名字中间会有 “-” 吗？

File: pkg/exporter/singleBridge.go [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+
+	s.OfSwitch = nil
+}
+
+func (s *SingleBridge) IsSwitchConnected() bool {
+	s.singleSwitchStatusMutex.Lock()
+	defer s.singleSwitchStatusMutex.Unlock()
+
+	return s.isSingleSwitchConnected
+}
```
Comment (KeyongSun) [inline pkg/exporter/singleBridge.go:84]: 这几个函数有地方调用和使用吗？

File: pkg/exporter/ovs.go (hunk 16 above)
Comment (cl1111) [reply pkg/exporter/ovs.go:122]: 在policyKeyWord/clsKeyWord/uplinkKeyWord中，增加了“-”

File: pkg/exporter/singleBridge.go (hunk 17 above)
Comment (cl1111) [reply pkg/exporter/singleBridge.go:84]: image
image
image
SwitchConnected和SwitchDisconnected是ofctl.AppInterface接口类型的方法，创建singleBridge的控制器时，需要将singleBridge实现为ofctl.AppInterface接口类型，所以需要实现SwitchConnected和SwitchDisconnected这两个方法。IsSwitchConnected后面新增加了调用。

File: pkg/exporter/ovs.go (hunk 12 above)
Comment (wu-cl) [reply pkg/exporter/ovs.go]: change "policy" to datapath.POLICY_BRIDGE_KEYWORD, etc.

Comment (cl1111) [reply pkg/exporter/ovs.go]: constant keywords are added to the current file
image
image

Comment (wu-cl) [reply pkg/exporter/ovs.go]: ok

**Review by KeyongSun — APPROVED** [review]

**Pull request merged**